### PR TITLE
rsa: Adds a signer for PSS and PKCS1#1.5

### PR DESCRIPTION
**Pull request by baloo**
This adds a mockhsm implementation and provides an integration test for both signers.

This was tested with `usb` against an hardware yubihsm.

### CODE REVIEW

File: src/client.rs [hunk 1]
```diff
@@ -1019,64 +1016,75 @@ impl Client {
...
     /// <https://developers.yubico.com/YubiHSM2/Commands/Sign_Pss.html>
-    #[cfg(feature = "untested")]
-    pub fn sign_rsa_pss_sha256(
+    pub(crate) fn sign_rsa_pss<S: SignatureAlgorithm>(
         &self,
         key_id: object::Id,
         data: &[u8],
     ) -> Result<rsa::pss::Signature, Error> {
         ensure!(
-            data.len() > rsa::pss::MAX_MESSAGE_SIZE,
```
Comment (baloo) [inline src/client.rs:1057]: I think this test was backwards.

The data should be `< 0xFFFF` as the data length is encoded in u16 as a prefix for the hash.

File: src/client.rs [hunk 2]
```diff
@@ -1019,64 +1016,75 @@ impl Client {
...
+            data.len() < rsa::pss::MAX_MESSAGE_SIZE,
             ErrorKind::ProtocolError,
             "message too large to be signed (max: {})",
             rsa::pss::MAX_MESSAGE_SIZE
         );
 
-        let mut hasher = Sha256::default();
-
-        let length = data.len() as u16;
-        hasher.update(length.to_be_bytes());
```
Comment (baloo) [inline src/client.rs:1066]: to note: I dropped this. I couldn't find where it was coming from, but that was producing incorrect signature.

Comment (baloo) [reply src/client.rs:1066]: (and this is a breaking change, but ... this is fixing signatures?)

File: src/asymmetric/algorithm.rs [hunk 3]
```diff
@@ -89,6 +89,14 @@ impl Algorithm {
             Algorithm::EcBp512 => 64,
         }
     }
+
+    /// Return the size of the given key (as expected by the `YubiHSM 2`) in bytes
```
Comment (tony-iqlusion) [inline src/asymmetric/algorithm.rs:93]: Comment seems wrong?

Comment (baloo) [reply src/asymmetric/algorithm.rs:93]: a lame copy paste from the `fn key_len`

**Comment by baloo** [comment]
I'll actually make this one a draft and wait for https://github.com/iqlusioninc/yubihsm.rs/pull/504

File: src/asymmetric/commands/put_key.rs [hunk 4]
```diff
@@ -33,3 +39,80 @@ pub(crate) struct PutAsymmetricKeyResponse {
...
+                    Some(v) => v,
+                    None => {
+                        return Err(de::Error::invalid_length(
+                            1usize,
+                            &"struct PutAsymmetricKeyCommand with 3 elements",
+                        ));
+                    }
+                };
+
+                let extra_data = if params.algorithm.rsa().is_some() {
```
Comment (baloo) [inline src/asymmetric/commands/put_key.rs:84]: I need a custom deserializer just for this reason.

**Comment by baloo** [comment]
both signers will need `signature::Keypair` and `spki::AssociatedAlgorithmIdentifier`

**Review by tony-iqlusion — APPROVED** [review]

**Pull request merged**